prow/pjutil/filter.go

```diff
@@ -97,7 +97,8 @@ func FilterPresubmits(filter Filter, changes config.ChangedFilesProvider, branch
 // RetestFilter builds a filter for `/retest`
 func RetestFilter(failedContexts, allContexts sets.String) Filter {
 	return func(p config.Presubmit) (bool, bool, bool) {
-		return failedContexts.Has(p.Context) || (!p.NeedsExplicitTrigger() && !allContexts.Has(p.Context)), false, true
+		failed := failedContexts.Has(p.Context)
+		return failed || (!p.NeedsExplicitTrigger() && !allContexts.Has(p.Context)), failed, true // force the trigger for failed jobs
```

Comment:
The final return value is
```
//  - what the default behavior should be if the presubmit
//    runs conditionally and does not match trigger conditions
```
So I would have expected returning `true` as the last return value would cause the jobs to run if the filter matches (first return value) even if the trigger conditions are not met which could point to a bug elsewhere. @stevekuznetsov does that sound right?

Reply:
Oh, I see what you mean! I misunderstood the code a bit, you are right. The case we were observing was indeed a job with `run_if_changed`, where *that* criteria were ruling the job out. If it were a job without `run_if_changed`, it would be triggered even with `always_run: false`, I guess.

In that case, the bug would be here I think:

https://github.com/kubernetes/test-infra/blob/ad24686e38736c3d111a07d0a96dd1291f2a0947/prow/config/jobs.go#L357-L362

The change regex matcher will explicity rule that job out because the job has a regex which does not match. The `default` will only apply when `determined==false`, which will only happen when the job does not have `run_if_changed`.

Reply:
So... `return shouldRun || default` ?

Reply:
I looked through this again and that seems right to me.

Reply:
Cool, I'll submit that change.